src/senaite/core/workflow/__init__.py

```diff
@@ -18,10 +18,214 @@
+    wf_tool = api.get_tool("portal_workflow")
+    workflow = wf_tool.getWorkflowById(workflow_or_id)
+    if not workflow:
+        if default is _marker:
+            raise ValueError("Workflow not found: %s " % workflow_or_id)
+        return default
+
+    return workflow
+
+
+def get_workflow_state(workflow_or_id, state_id, default=_marker):
+    """Returns the workflow state with the given id
+    """
+    wf = get_workflow(workflow_or_id)
+    state = wf.states.get(state_id)
+    if not state:
+        if default is _marker:
+            raise ValueError("State %s not found for %s" % (state_id, wf.id))
+        return default
+    return state
+
+
+def get_workflow_transition(workflow_or_id, transition_id, default=_marker):
+    """Returns the workflow transition with the given id
+    """
+    wf = get_workflow(workflow_or_id)
+    transition = wf.transitions.get(transition_id)
+    if not transition:
+        if default is _marker:
+            raise ValueError("Transition %s not found for %s" %
+                             (transition_id, wf.id))
+        return default
+    return transition
+
+
+def update_workflow(workflow, **kwargs):
+    """Updates the workflow with workflow_id with the settings passed-in
+    """
+    wf = get_workflow(workflow)
+
+    # Set basic info (title, description, etc.)
+    wf.title = kwargs.get("title", wf.title)
+    wf.description = kwargs.get("description", wf.description)
+    wf.initial_state = kwargs.get("initial_state", wf.initial_state)
+
+    # Update states
+    states = kwargs.get("states", {})
+    for state_id, values in states.items():
+
+        # Create the state if it does not exist yet
+        if not wf.states.get(state_id):
+            wf.states.addState(state_id)
+
+        # Update the state with the settings passed-in
+        update_workflow_state(wf, state_id, **values)
+
+    # Update transitions
+    transitions = kwargs.get("transitions", {})
+    for transition_id, values in transitions.items():
+
+        # Create the transition if it does not exist yet
+        if not wf.transitions.get(transition_id):
+            wf.transitions.addTransition(transition_id)
+
+        # Update the transition with the settings passed-in
+        update_workflow_transition(wf, transition_id, **values)
+
+
+def update_workflow_state(workflow, state_id, **kwargs):
+    """Updates the status of a workflow in accordance with settings passed-in
+
+    Note that regarding the use of tuples and lists for permissions and
+    transitions, the same principles from DCWorkflow apply. This is:
+
+    - Transitions passed-in as a list are added to existing transitions
+    - Transitions passed-in as a tuple replace the existing transitions
+    """
+    wf = get_workflow(workflow)
+    state = get_workflow_state(wf, state_id)
+
+    # Set basic info (title, description, etc.)
+    state.title = kwargs.get("title", state.title)
+    state.description = kwargs.get("description", state.description)
```

Comment:
Initially I thought to pass in parameters in the following way:

`update_workflow_state(workflow, state_id, title=None, description=None, states=None, transitions=None, **kwargs)`

where the `states` and `transitions` would be the dicts with the parameters.